WordPress/gutenberg
# Query and Search blocks: support for Instant Search via `query_loop_block_query_vars` filter

## What?
Implementation of Instant Search using the Search and Query Loop blocks. Added as a new experiment on the Gutenberg Experiments page.

Related to [#63053](https://github.com/WordPress/gutenberg/issues/63053)

https://github.com/user-attachments/assets/7ee8ddc7-472c-4575-a27e-af1972278c6c


## How does it work?

### Testing

1. Enable the `Instant Search and Query Block` Gutenberg experiment.
2. Insert a Query block anywhere on your site. 
3. Make sure to disable `Reload Full Page` checkbox in the Query block's Advanced Settings i.e. use "enhanced pagination" in that Query block.
4. Insert a Search block anywhere inside of that Query block.

When the Search block is inside of the Query block using "enhanced pagination", it automatically gets "updated" to an Instant Search block. 

Any search input in the Instant Search block gets passed as the `?instant-search-<queryId>=<search-term>` query search param in the URL.

### Multiple Query + Search blocks

It's possible to have multiple Instant Search blocks in a page/post/template. In such a case, ensuring their functionality is isolated is essential. For this to work, the `queryId` is part of the param syntax:
- `instant-search-<queryId>=<search-term>`

### Search button

The search block can contain a search button (in some configurations). However, when using the Instant Search functionality, the button is redundant because the list of posts updates as you type. For this reason, **in the editor**, when the Search block is placed inside the Query Loop block with enhanced pagination ON, the Block Controls related to the Search button are removed. The displayed name of the block (via `label`) is changed to `Instant Search (Search)`.

**On the frontend**, the Search button is also always removed for each (Instant) Search block.


https://github.com/user-attachments/assets/f75ba0a5-2672-4d36-839b-253f25fa4de8


### Pagination

The instant search functionality **respects** the pagination of the Query block, which uses the `query-<queryId>-page` syntax.

1. Every time the search is updated, the pagination of its respective query is reset back to page 1. 
2. The pagination numbers and next/previous, are updated to reflect the number of results in the search.
3. Clearing the search also resets the pagination back to page 1.

### Limitations

- Only the `Custom` query types are supported in the current experiment. The `Default` query types are currently not supported but will be in the future version. This is prototyped already in https://github.com/WordPress/gutenberg/pull/67289.

## Alternatives

Alternative approaches have been explored: 

- https://github.com/WordPress/gutenberg/pull/63147- modifying the server-rendering of the individual blocks (post-template, query-pagination, etc.)
- https://github.com/WordPress/gutenberg/pull/67013 - using the `render_block_context` filter.


## Further work

This is an initial prototype and the following features is **intentionally** not yet implemented:
- Support for the `Default` queries, including multiple `Default` queries on a single page/post/template. Those have been prototyped already in https://github.com/WordPress/gutenberg/pull/67289.
- Explore an alternative mechanism to convert the Search block into a Instant Search. Currently, this is done by nesting the Search inside of Query Loop with enhanced pagination ON, but a different, more explicit mechanism might be appropriate.
- Include contents of the Query No Results block but hide them with CSS ([comment](https://github.com/WordPress/gutenberg/pull/63147#discussion_r1860951647))
